primitives/trie/src/cache/shared_cache.rs

```diff
@@ -0,0 +1,632 @@
+			// value to early.
+			self.lru.get(&ValueCacheKey::Hash {
+				hash: key,
+				_i_read_the_documentation: IReadTheDocumentation(()),
+			});
+		});
+
+		added.into_iter().for_each(|(key, value)| {
+			let (storage_root, storage_key, key_hash) = match key {
+				ValueCacheKey::Hash { .. } => {
+					// Ignore the hash variant and try the next.
+					tracing::error!(
+						target: LOG_TARGET,
+						"`SharedValueCached::update` was called with a key to add \
+						that uses the `Hash` variant. This would lead to potential hash collision!",
+					);
+					return
+				},
+				ValueCacheKey::Ref { storage_key, storage_root, hash } =>
+					(storage_root, storage_key.into(), hash),
+				ValueCacheKey::Value { storage_root, storage_key, hash } =>
+					(storage_root, storage_key, hash),
+			};
+
+			let (size_update, storage_key) =
+				match self.known_storage_keys.entry(storage_key.clone()) {
+					SetEntry::Vacant(v) => {
+						let len = v.get().len();
+						v.insert();
+
+						// If the key was unknown, we need to also take its length and the size of
+						// the entry of `known_keys` into account.
+						(len + base_size + known_keys_entry_size, storage_key)
+					},
+					SetEntry::Occupied(o) => {
+						// Key is known
+						(base_size, o.get().clone())
+					},
+				};
+
+			self.size_in_bytes += size_update;
+
+			if let Some((r_key, _)) = self
+				.lru
+				.push(ValueCacheKey::Value { storage_key, storage_root, hash: key_hash }, value)
+			{
+				if let ValueCacheKey::Value { storage_key, .. } = r_key {
+					update_size_in_bytes(
+						&mut self.size_in_bytes,
+						storage_key,
+						&mut self.known_storage_keys,
+					);
+				}
+			}
+
+			// Directly ensure that we respect the maximum size. By doing it directly here we
+			// ensure that the internal map of the [`LruCache`] doesn't grow too much.
+			while self.maximum_cache_size.exceeds(self.size_in_bytes) {
+				// This should always be `Some(_)`, otherwise something is wrong!
+				if let Some((r_key, _)) = self.lru.pop_lru() {
+					if let ValueCacheKey::Value { storage_key, .. } = r_key {
+						update_size_in_bytes(
+							&mut self.size_in_bytes,
+							storage_key,
+							&mut self.known_storage_keys,
+						);
+					}
+				}
+			}
+		});
+	}
+
+	/// Reset the cache.
+	fn reset(&mut self) {
+		self.size_in_bytes = 0;
+		self.lru.clear();
+		self.known_storage_keys.clear();
+	}
+}
+
+/// The shared trie cache.
+///
+/// It should be instantiated once per node. It will hold the trie nodes and values of all
+/// operations to the state. To not use all available memory it will ensure to stay in the
+/// bounds given via the [`CacheSize`] at startup.
+///
+/// The instance of this object can be shared between multiple threads.
+pub struct SharedTrieCache<H: Hasher> {
+	pub(super) node_cache: Arc<RwLock<SharedNodeCache<H>>>,
+	pub(super) value_cache: Arc<RwLock<SharedValueCache<H::Out>>>,
```

Comment:
Locking order for these is not well defined. There seems to be only one place where both locks are taken simultaneously, so that's probably not an issue. But I'd add a comment nevertheless.